landlab/components/lake_fill/lake_fill_barnes.py

```diff
@@ -0,0 +1,2149 @@
+        >>> z_init = z.copy()
+        >>> lmb = LakeMapperBarnes(mg, method='steepest', surface=z_init,
+        ...                        fill_flat=False,
+        ...                        redirect_flow_steepest_descent=False,
+        ...                        track_lakes=False)
+
+        TODO: once return_array_at_node is fixed, this example should also
+        take fill_surface...
+
+        >>> lmb.run_one_step()
+        >>> z_out = np.array([ 0. ,  0. ,  0. ,  0. ,  0. ,  0. ,
+        ...                    0. ,  2.1,  1.5,  1.5,  1.6,  0. ,
+        ...                    0. ,  2. ,  1.5,  1.5,  1.5,  0. ,
+        ...                    0. ,  2.2,  1.5,  1.5,  1.7,  0. ,
+        ...                    0. ,  0. ,  0. ,  0. ,  0. ,  0. ])
+        >>> np.allclose(z, z_out)
+        True
+        >>> np.all(np.equal(z, z_out))  # those 1.5's are actually a bit > 1.5
+        False
+        >>> try:
+        ...     lmb.lake_map  # not created, as we aren't tracking
+        ... except AssertionError:
+        ...     print('AssertionError was raised: ' +
+        ...           'Enable tracking to access information about lakes')
+        AssertionError was raised: Enable tracking to access information about lakes
+        >>> lmb.was_there_overfill  # everything fine with slope adding
+        False
+
+        >>> fd = FlowDirectorSteepest(mg)
+        >>> fa = FlowAccumulator(mg)  # routing will work fine now
+        >>> fd.run_one_step()
+        >>> fa.run_one_step()
+        >>> np.all(mg.at_node['flow__sink_flag'][mg.core_nodes] == 0)
+        True
+        >>> drainage_area = np.array([  0.,   0.,   0.,   0.,   0.,   0.,
+        ...                             0.,   4.,   8.,  12.,   4.,   4.,
+        ...                             0.,   4.,  16.,  36.,  40.,  40.,
+        ...                             0.,   4.,   8.,   4.,   4.,   4.,
+        ...                             0.,   0.,   0.,   0.,   0.,   0.])
+        >>> np.allclose(mg.at_node['drainage_area'], drainage_area)
+        True
+
+        Test two pits:
+
+        >>> z[:] = mg.node_x.max() - mg.node_x
+        >>> z[23] = 1.3
+        >>> z[15] = 0.3
+        >>> z[10] = 1.3  # raise "guard" exit nodes
+        >>> z[7] = 2.  # is a lake on its own, if d8
+        >>> z[9] = 0.5
+        >>> z[14] = 0.6  # [9, 14, 15] is a lake in both methods
+        >>> z[16] = 1.2
+        >>> z[22] = 0.9  # a non-contiguous lake node also draining to 16 if d8
+        >>> z_init = z.copy()
+        >>> lmb = LakeMapperBarnes(mg, method='d8', fill_flat=True,
+        ...                        track_lakes=True)
+        >>> lmb.run_one_step()  # note the d8 routing now
+        >>> lmb.lake_dict == {22: deque([15, 9, 14])}
+        True
+        >>> lmb.number_of_lakes
+        1
+        >>> try:
+        ...     lmb.lake_depths  # z was both surface and 'fill_surface'
+        ... except AssertionError:
+        ...     print('AssertionError was raised: ' +
```

Comment:
@mcflugen , correct me if you think otherwise, but would it be better if these sorts of tests were done as unit tests instead of docstring tests? Does it matter?

Since this is a private function, these docstrings won't clutter up the documentation anyway. At least I think thats how we have sphinx set up.

Reply:
I quite want these to be visible to the user, since they might be a little counter-intuitive. Others going in unittests though.